data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,498 @@
+            query_job_ids.append(query_job_id)
+
+    incomplete_jobs = response.exception()
+    if incomplete_jobs is not None:
+        count = len(incomplete_jobs)
+        if incomplete_jobs:
+            LOGGER.info(f"Failed on {count} job ids {incomplete_jobs}")
+            LOGGER.info("Terminating retraction")
+            raise incomplete_jobs
+
+
+def parse_args(raw_args=None):
+    parser = argparse.ArgumentParser(
+        description=
+        'Runs retraction of deactivated EHR participants on all datasets in project. The rows to be retracted '
+        'are based on the date entered and will be retracted if the date entered is before the retraction. '
+        'Uses project_id, deactivated pid_table_id to determine '
+        'the pids to retract data for. The pid_table_id needs to contain '
+        'the person_id and deactivated_date columns specified in the schema above, '
+        'but research_id can be null if deid has not been run yet. ',
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('-p',
+                        '--project-id',
+                        action='store',
+                        dest='project_id',
+                        help='Identifies the project to retract data from',
+                        required=True)
+    parser.add_argument('-t',
+                        '--ticket-number',
+                        action='store',
+                        dest='ticket_number',
+                        help='Ticket number to append to sandbox table names',
+                        required=True)
+    parser.add_argument(
+        '-i',
+        '--pids-project-id',
+        action='store',
+        dest='pids_project_id',
+        help='Identifies the project where the pids table is stored',
+        required=True)
+    parser.add_argument(
+        '-d',
+        '--pids-dataset-id',
+        action='store',
+        dest='pids_dataset_id',
+        help='Identifies the dataset where the pids table is stored',
+        required=True)
+    parser.add_argument('-s',
+                        '--pids-table',
+                        action='store',
+                        dest='pids_table',
+                        help='Identifies the table where the pids are stored',
+                        required=True)
+    parser.add_argument('-c',
+                        '--console-log',
+                        dest='console_log',
+                        action='store_true',
+                        required=False,
+                        help='Log to the console as well as to a file.')
+    return parser.parse_args(raw_args)
+
+
+def main(args=None):
+    args = parse_args(args)
+    add_console_logging(args.console_log)
+    query_list = create_queries(args.project_id, args.ticket_number,
+                                args.pids_project_id, args.pids_dataset_id,
+                                args.pids_table)
+    client = bq.get_client(args.project_id)
```

Comment:
this module uses a lot of queries just to generate the retraction queries.  I would move this before `create_queries` and either pass this client to the `create_queries` function or create the other client creation method listed in a different comment.  Admittedly, I didn't see this first.  I've been following the module dataflow.

Reply:
passing client to `create_queries` I don't think is necessary since it's hitting the `bq.py` to create the client. Which if I refactor `bq.py` this will solve the multiple calls. I am using client in `run_queries` to actually execute the retractions and sandboxing: `response = client.query(query_dict['query'])`

Reply:
I now see what you mean, adjusting this! I will call client once before query_list and pass for all other functions to use.